state.go

```diff
@@ -27,76 +31,804 @@ type State struct {
+	}
+
+	s.Keys = *fe.Next(leafCount(s.Tree.size()), updateSecret, encGrpCtx)
+
+	// confirmation verification
+	hmac := suite.newHMAC(s.Keys.ConfirmationKey)
+	hmac.Write(s.ConfirmedTranscriptHash)
+	confirm := hmac.Sum(nil)
+	if !bytes.Equal(confirm, gi.Confirmation) {
+		return nil, fmt.Errorf("mls.state: confirmation failed to verif")
+	}
+
+	return s, nil
 }
 
 func negotiateWithPeer(groupID []byte, myCIKs, otherCIKs []ClientInitKey, commitSecret []byte) (*Welcome, *State, error) {
-	// TODO
-	return nil, nil, nil
+	var selected = false
+	var mySelectedCik, otherSelectedCik ClientInitKey
+
+	for _, mycik := range myCIKs {
+		for _, ocik := range otherCIKs {
+			if mycik.CipherSuite == ocik.CipherSuite {
+				selected = true
+				mySelectedCik = mycik
+				otherSelectedCik = ocik
+				break
+			}
+		}
+		if selected {
+			break
+		}
+	}
+
+	if !selected {
+		return nil, nil, fmt.Errorf("mls.state: negotiation failure")
+	}
+
+	// init our state and add the negotiated peer's cik
+	s := newEmptyState(groupID, mySelectedCik.CipherSuite, *mySelectedCik.privateKey, mySelectedCik.Credential)
+	add := s.add(otherSelectedCik)
+	// update tree state
+	_, err := s.handle(add)
+	if err != nil {
+		return nil, nil, err
+	}
+	// commit the add and generate welcome to be sent to the peer
+	_, welcome, newState, err := s.commit(commitSecret)
+	if err != nil {
+		panic(fmt.Errorf("mls.state: commit failure"))
+	}
+
+	return welcome, newState, nil
+}
+
+func (s State) add(cik ClientInitKey) *MLSPlaintext {
+	addProposal := Proposal{
+		Add: &AddProposal{
+			ClientInitKey: cik,
+		},
+	}
+	return s.sign(addProposal)
+}
+
+func (s State) update(leafSecret []byte) *MLSPlaintext {
+	key, err := s.CipherSuite.hpke().Derive(leafSecret)
+	if err != nil {
+		panic(fmt.Errorf("mls.state: deriving secret for update failure %v", err))
+	}
+
+	updateProposal := Proposal{
+		Update: &UpdateProposal{
+			LeafKey: key.PublicKey,
+		},
+	}
+
+	return s.sign(updateProposal)
+}
+
+func (s *State) remove(removed leafIndex) *MLSPlaintext {
+	removeProposal := Proposal{
+		Remove: &RemoveProposal{
+			Removed: uint32(removed),
+		},
+	}
+	return s.sign(removeProposal)
+}
+
+func (s *State) commit(leafSecret []byte) (*MLSPlaintext, *Welcome, *State, error) {
+	commit := Commit{}
+	var joiners []ClientInitKey
+
+	for _, pp := range s.PendingProposals {
+		pid := s.proposalId(pp)
+		if s.processedProposals[pid.String()] {
+			fmt.Printf("mls.state: skipping processed proposal %v", pp)
+			continue
+		}
+
+		proposal := pp.Content.Proposal
+		switch {
+		case proposal.Add != nil:
+			commit.Adds = append(commit.Adds, pid)
+			joiners = append(joiners, proposal.Add.ClientInitKey)
+		case proposal.Update != nil:
+			commit.Updates = append(commit.Updates, pid)
+		case proposal.Remove != nil:
+			commit.Removes = append(commit.Removes, pid)
+		}
+	}
+
+	// init new state to apply commit and ratchet forward
+	next := s.clone()
+	next.apply(commit)
+	next.PendingProposals = nil
```

Comment:
I would just do this inside of clone(), i.e., don't clone the pending proposals.  That might mean it would be better to rename "clone" to something like "successor"